### PR TITLE
implement parsing of cosmos-sdk/metro transactions and namespacing

### LINKED ISSUE #4: implement namespacing of transactions
Transactions belonging to different rollups should be stored under different namespaces. The relayer should parse each block for the transactions for each rollup and write those together. It should also write a "sequencer block" that contains data that points to all the different rollup transaction batch namespaces, so that the entire block (will all rollup txs) can be retrieved.

Questions:
- how are namespaces registered for each rollup? how to we ensure no conflicts?
- what's the ideal format for the data that points to each of the rollup's data?

### PR DESCRIPTION
**Pull request by noot**
- implement parsing of cosmos-sdk/metro transactions via protobuf
- implement parsing a metro block into different namespaces and write to namespaces separately 

closes #4

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -8,11 +8,17 @@ anyhow = "1.0"
 async-trait = "0.1"
 base64 = "0.21"
 hex = "0.4"
+lazy_static = "1.4"
+protobuf = { version = "2.28", features = ["with-bytes"] }
```
[the remark this answers is not in the document]
Comment (noot) [reply Cargo.toml:12]: I hadn't seen prost before! do you have one you'd recommend?

File: src/bin/relayer.rs [hunk 2]
```diff
@@ -72,8 +73,18 @@ async fn main() {
 
                 info!("got block with height {} from sequencer", height);
                 highest_block_number = height;
-                let tx_count = resp.block.data.txs.len();
-                match da_client.submit_block(resp.block.into()).await {
+                let sequencer_block = match SequencerBlock::from_cosmos_block(resp.block) {
+                    Ok(block) => block,
+                    Err(e) => {
+                        warn!("failed to convert block to DA block: {:?}", e);
+                        thread::sleep(sleep_duration);
```
[the remark this answers is not in the document]
Comment (noot) [reply src/bin/relayer.rs:80]: yeah, I think we should sleep here - there isn't really a point in constantly querying for the next block, when we already know what the block time is

File: src/da.rs [hunk 3]
```diff
@@ -1,50 +1,36 @@
...
-    txs: Vec<Base64String>,
-}
 
-impl SequencerBlock {
-    pub fn new() -> Self {
-        SequencerBlock { txs: vec![] }
-    }
+lazy_static! {
+    #[derive(Hash, Eq, PartialEq, Debug)]
+    static ref DEFAULT_NAMESPACE: String = get_namespace(b"sequencer");
```
[the remark this answers is not in the document]
Comment (noot) [reply src/da.rs:17]: this requires nightly rust I believe :( it looks like a good solution though. do you think it's worth moving to nightly for this? (or should I just go with the tokio OnceCell suggestion?)

File: Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (steezeburger) [reply Cargo.toml:12]: We've also talked about using https://buf.build/ for our RPC stuff. Their CLI has a plugin that uses Prost under the hood.

You can see the public rpc repo that I've done so far here:
https://buf.build/astria/execution-apis

File: src/sequencer_block.rs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+fn parse_cosmos_tx(tx: &Base64String) -> Result<TxBody, Error> {
+    let tx_raw = TxRaw::parse_from_bytes(&tx.0)?;
+    let tx_body = TxBody::parse_from_bytes(&tx_raw.body_bytes)?;
+    Ok(tx_body)
+}
+
+fn cosmos_tx_body_to_sequencer_msgs(tx_body: TxBody) -> Result<Vec<SequencerMsg>, Error> {
+    let mut msgs = vec![];
+    for msg in tx_body.messages {
```
[the remark this answers is not in the document]
Comment (noot) [reply src/sequencer_block.rs:87]: I tried implementing this but the resulting type isn't a `Result` :( It's a `Vec<Result<SequencerMsg, Error>>` and I'm not totally sure how to cleanly turn that into a `Result<Vec<SequencerMsg>, Error>` like before

[the remark this answers is not in the document]
Comment (noot) [reply src/sequencer_block.rs]: I got it working :D

File: src/da.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (noot) [reply src/da.rs:17]: I tried using your `sha2-const` fork, which allows me to do this:
```
const DEFAULT_NAMESPACE: [u8; 32] = Sha256::new()
    .update(
        b"sequencer",
    )
    .finalize();
```

but then for every usage of `DEFAULT_NAMESPACE`, I need to do `hex::encode(DEFAULT_NAMESPACE[0..8])` :( I wish there was a way to slice and encode it in a constant. will try `OnceCell` for now

File: src/da.rs [hunk 5]
```diff
@@ -1,50 +1,36 @@
...
-#[derive(Serialize, Deserialize, Debug, Default)]
-pub struct SequencerBlock {
-    txs: Vec<Base64String>,
-}
 
-impl SequencerBlock {
-    pub fn new() -> Self {
-        SequencerBlock { txs: vec![] }
-    }
+lazy_static! {
```
[the remark this answers is not in the document]
Comment (noot) [reply src/da.rs:15]: do you think I need to tokio OnceCell, or is the nonasync OnceCell okay? with the tokio OnceCell I end up writing:
```
static DEFAULT_NAMESPACE: OnceCell<String> = OnceCell::const_new();

async fn get_default_namespace() -> String {
    DEFAULT_NAMESPACE.get_or_init(|| async { get_namespace(b"sequencer") }).await.to_string()
}
```

`get_namespace` isn't async, so this seems a bit unnecessary lol

[the remark this answers is not in the document]
Comment (noot) [reply src/da.rs:15]: it's also kind of unclear to me when to call `get_default_namespace()` in tests for example, do I just call it at the start of every test, and also in `main()`?

File: Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (noot) [reply Cargo.toml]: cool! I updated to use prost.

File: src/da.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (noot) [reply src/da.rs]: I ended up implementing a `Namespace` type, similar to your second example. I think that's the nicest!

**Pull request merged**